src/main/java/org/eclipsefoundation/marketplace/service/impl/DefaultPromotionService.java

```diff
@@ -0,0 +1,171 @@
+import org.eclipsefoundation.marketplace.dto.Listing;
+import org.eclipsefoundation.marketplace.dto.Promotion;
+import org.eclipsefoundation.marketplace.dto.filter.DtoFilter;
+import org.eclipsefoundation.marketplace.helper.StreamHelper;
+import org.eclipsefoundation.marketplace.model.MongoQuery;
+import org.eclipsefoundation.marketplace.model.RequestWrapper;
+import org.eclipsefoundation.marketplace.namespace.MicroprofilePropertyNames;
+import org.eclipsefoundation.marketplace.namespace.UrlParameterNames;
+import org.eclipsefoundation.marketplace.service.CachingService;
+import org.eclipsefoundation.marketplace.service.PromotionService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Default implementation of promotion service. Uses weighting to allow for
+ * promotions to appear more often than others. By using the property
+ * {@link MicroprofilePropertyNames.PROMO_WEIGHT_DEFAULT}, weighting defaults
+ * can shift outside of code builds once data is modified.
+ * 
+ * @author Martin Lowe
+ *
+ */
+@ApplicationScoped
+public class DefaultPromotionService implements PromotionService {
+	private static final Logger LOGGER = LoggerFactory.getLogger(DefaultPromotionService.class);
+
+	@ConfigProperty(name = MicroprofilePropertyNames.PROMO_SERVE_COUNT, defaultValue = "2")
+	int promoCount;
+	@ConfigProperty(name = MicroprofilePropertyNames.PROMO_WEIGHT_DEFAULT, defaultValue = "1")
+	int defaultWeight;
+
+	@Inject
+	MongoDao dao;
+
+	@Inject
+	DtoFilter<Listing> listingFilter;
+	@Inject
+	CachingService<List<Listing>> listingCache;
+
+	@Inject
+	DtoFilter<Promotion> promotionFilter;
+	@Inject
+	CachingService<List<Promotion>> promoCache;
+
+	// random used for shuffling collections
+	private Random r = new Random();
+
+	@Override
+	public List<Listing> getListingsForPromotions(RequestWrapper wrapper, List<Promotion> promos) {
+		if (promos == null || promos.isEmpty()) {
+			LOGGER.debug("No promotions were passed, returning empty list");
+			return Collections.emptyList();
+		}
+		// create mapping to get a list of specific IDs, and to add context to the call
+		// for caching
+		Map<String, List<String>> adds = new HashMap<>();
+		adds.put("type", Arrays.asList("Listing"));
+		adds.put(UrlParameterNames.IDS.getParameterName(),
+				promos.stream().map(Promotion::getListingId).collect(Collectors.toList()));
+
+		MongoQuery<Listing> q = new MongoQuery<>(null, adds, listingFilter);
+		// retrieve the possible cached object
+		Optional<List<Listing>> cachedResults = listingCache.get("promo|listings", wrapper, adds,
+				() -> StreamHelper.awaitCompletionStage(dao.get(q)));
+		if (!cachedResults.isPresent()) {
+			LOGGER.error("Error while retrieving cached promotion listings");
+			return Collections.emptyList();
+		}
+
+		// return the results as a response
+		return cachedResults.get();
+	}
+
+	@Override
+	public void injectPromotions(RequestWrapper wrapper, List<Listing> listings) {
```

Comment:
do not modifiy list given in parameters. Returns a new list and if needed, when you call the injectPromotions method, do a addAll to the list you were previously passing as parameter. It's more testable.